### ISSUE TITLE
Linux: Support for systemd-resolved is absent

### ISSUE THREAD

**Issue by RogueScholar** [mullvad/mullvadvpn-app#495]
I have several systems running Ubuntu or Ubuntu derivatives on which the Mullvad app fails to function. All of the systems use systemd-resolved for DNS resolution (the new default resolver which replaces resolvconf in 18.04 Bionic Beaver and beyond) and the underlying issue appears to be a failure to recognize when /usr/bin/resolvconf is merely a symlink to /usr/bin/resolvectl (the multi-call binary which provides CLI access to systemd-resolved and a compatibility mode when called as resolvconf). The error shown in the logs is:

> [talpid_core::security][INFO] Applying security policy: Connected to [REDACTED]:1301:UDP over "tun0" (ip: [REDACTED], gw: [REDACTED]), Allowing LAN
> [talpid_core::tunnel_state_machine::connected_state][ERROR] Error: Failed to apply security policy for connected state
> Caused by: Failed to execute 'resolvconf' program
> Caused by: command ["/usr/bin/resolvconf", "-a", "tun0.mullvad"] exited with code 1

It would seem that either the app needs to recognize when systemd-resolved is acting in place of resolvconf and adjust its syntax to satisfy the compatibility mode, or as advised by systemd documentation, instead use DBus to signal the desired changes to the DNS resolution configuration. I use this script, [update-systemd-resolved](https://github.com/jonathanio/update-systemd-resolved) in my OpenVPN config files to accomplish that rather painlessly. The end of each config file ends up looking like so:

```
dhcp-option DOMAIN-ROUTE .

script-security 2
setenv PATH /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
up /etc/openvpn/scripts/update-systemd-resolved
down /etc/openvpn/scripts/update-systemd-resolved
down-pre
```

The new app is really impressive and I have nothing but good experiences to report with it on my Windows boxes. I can't wait to start using it on my Linux machines as well.

**Comment by faern** [comment]
Thanks for the report. This is already being worked on in #489

**Comment by RogueScholar** [comment]
Installed 2018.4-beta2 today to examine the new systemd-resolved support and was treated to a painless connection experience with all GUI functions appearing to work like a charm. You guys are rock stars!

I did observe one remaining issue, though, and that's the absence of a link domain entry for the OpenVPN interface, leading to a predictable DNS leak. It's rather painlessly corrected by issuing the following command in the terminal

`resolvectl domain tun0 ~.`

The optimal way to set it though is from within the openvpn binary, at least as far as I can see. Referring back to my earlier post I was doing that with this line added to openvpn-client.conf:

`dhcp-option DOMAIN-ROUTE .`

I believe that same flag needs to be in the output of the routine where you build your openvpn expression, but of course you would know best. Thanks again for making such quick work of this issue, and allowing me to learn a thing or two by watching your work.

**Comment by pinkisemils** [comment]
Yes, in 2018.4 we did not set the link domain entry, and this indeed could lead to a DNS leak. We have fixed this in master, and the upcoming beta should have the fixes in place (#553).

We actually don't rely on OpenVPN to adjust any system config aside from setting routes. This is mostly because we want to target as many platforms and distributions as possible. Hopefully, DNS issues will be solved for good now. But we're always open to hearing about more issues :slightly_smiling_face:
